### PR TITLE
feat(zap): add multiple zapcore support

### PR DESCRIPTION
**Pull request by StellarisW**
#### What type of PR is this?

feat

#### What this PR does / why we need it (English/Chinese):

目前`hlog`的`zap`扩展不支持使用多个`zapcore`的场景，比如我既需要输出到控制台，又需要输出到日志文件

### CODE REVIEW

**Comment by rogerogers** [comment]
```go
logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
	return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
})))
```
maybe we could use zap.WrapCore custom our zapcore?

**Comment by StellarisW** [comment]
> ```go
> 	logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> 		return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> 	})))
> ```
> 
> maybe we could use zap.WrapCore custom our zapcore?

`WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless

**Comment by rogerogers** [comment]
```go
l, err := os.OpenFile("log.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0755)
if err != nil {
	log.Fatal(err)
}
logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
	return zapcore.NewTee([]zapcore.Core{
		zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(l), zap.NewAtomicLevelAt(zap.InfoLevel)),
		zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.InfoLevel)),
	}...)
})))
```
multiple zapcore by NewTee

**Comment by StellarisW** [comment]
> ```go
> l, err := os.OpenFile("log.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0755)
> if err != nil {
> 	log.Fatal(err)
> }
> logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> 	return zapcore.NewTee([]zapcore.Core{
> 		zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(l), zap.NewAtomicLevelAt(zap.InfoLevel)),
> 		zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.InfoLevel)),
> 	}...)
> })))
> ```
> 
> multiple zapcore by NewTee

it indeed a good way to use multiple zapcores in this way

**Comment by rogerogers** [comment]
> > ```go
> > l, err := os.OpenFile("log.log", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0755)
> > if err != nil {
> > 	log.Fatal(err)
> > }
> > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> > 	return zapcore.NewTee([]zapcore.Core{
> > 		zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(l), zap.NewAtomicLevelAt(zap.InfoLevel)),
> > 		zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.InfoLevel)),
> > 	}...)
> > })))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > multiple zapcore by NewTee
> 
> it indeed a good way to use multiple zapcores in this way

We can mention this in our documentation, the code change is not necessary for this repository

**Comment by rogerogers** [comment]
> > ```go
> > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> >     return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> > })))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > maybe we could use zap.WrapCore custom our zapcore?
> 
> `WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless

WrapCore will take effect finally, and will cause setOutput and setLevel not to work

**Comment by StellarisW** [comment]
> > > ```go
> > > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> > >     return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> > > })))
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > maybe we could use zap.WrapCore custom our zapcore?
> > 
> > 
> > `WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless
> 
> WrapCore will take effect finally, and will cause setOutput and setLevel not to work

In this case, to solve the warped `zapcore.Core` probles, we still need change methods `SetLevel` and `SetOutput` by `NewTee` and it will just change the first `zapcore.Core` in the warped `zapcore.Core`

**Comment by rogerogers** [comment]
> > > > ```go
> > > > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> > > >     return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> > > > })))
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > maybe we could use zap.WrapCore custom our zapcore?
> > > 
> > > 
> > > `WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless
> > 
> > 
> > WrapCore will take effect finally, and will cause setOutput and setLevel not to work
> 
> In this case, to solve the warped `zapcore.Core` probles, we still need change methods `SetLevel` and `SetOutput` by `NewTee` and it will just change the first `zapcore.Core` in the warped `zapcore.Core`

It will look weird, and it's not so easy to change the wrapCore behavior, maybe we should avoid using these two func when using wrapCore

**Comment by StellarisW** [comment]
> > > > > ```go
> > > > > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> > > > >     return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> > > > > })))
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > maybe we could use zap.WrapCore custom our zapcore?
> > > > 
> > > > 
> > > > `WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless
> > > 
> > > 
> > > WrapCore will take effect finally, and will cause setOutput and setLevel not to work
> > 
> > 
> > In this case, to solve the warped `zapcore.Core` probles, we still need change methods `SetLevel` and `SetOutput` by `NewTee` and it will just change the first `zapcore.Core` in the warped `zapcore.Core`
> 
> It will look weird, and it's not so easy to change the wrapCore behavior, maybe we should avoid using these two func when using wrapCore

... how's weeird?

**Comment by rogerogers** [comment]
> > > > > > ```go
> > > > > > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> > > > > >     return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> > > > > > })))
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > maybe we could use zap.WrapCore custom our zapcore?
> > > > > 
> > > > > 
> > > > > `WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless
> > > > 
> > > > 
> > > > WrapCore will take effect finally, and will cause setOutput and setLevel not to work
> > > 
> > > 
> > > In this case, to solve the warped `zapcore.Core` probles, we still need change methods `SetLevel` and `SetOutput` by `NewTee` and it will just change the first `zapcore.Core` in the warped `zapcore.Core`
> > 
> > 
> > It will look weird, and it's not so easy to change the wrapCore behavior, maybe we should avoid using these two func when using wrapCore
> 
> ... how's weeird?

Why just change the first...

**Comment by StellarisW** [comment]
> > > > > > > ```go
> > > > > > > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> > > > > > >     return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> > > > > > > })))
> > > > > > > ```
> > > > > > > 
> > > > > > > 
> > > > > > >     
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >     
> > > > > > >   
> > > > > > > maybe we could use zap.WrapCore custom our zapcore?
> > > > > > 
> > > > > > 
> > > > > > `WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless
> > > > > 
> > > > > 
> > > > > WrapCore will take effect finally, and will cause setOutput and setLevel not to work
> > > > 
> > > > 
> > > > In this case, to solve the warped `zapcore.Core` probles, we still need change methods `SetLevel` and `SetOutput` by `NewTee` and it will just change the first `zapcore.Core` in the warped `zapcore.Core`
> > > 
> > > 
> > > It will look weird, and it's not so easy to change the wrapCore behavior, maybe we should avoid using these two func when using wrapCore
> > 
> > 
> > ... how's weeird?
> 
> Why just change the first...

As It can't specify the certain `zapcore.Core`, we have to to implement `Control` interface (just point this case in docs is ok)

**Comment by rogerogers** [comment]
> > > > > > > > ```go
> > > > > > > > logger := hertzzap.NewLogger(hertzzap.WithZapOptions(zap.WrapCore(func(core zapcore.Core) zapcore.Core {
> > > > > > > >     return zapcore.NewCore(zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), zapcore.AddSync(os.Stdout), zap.NewAtomicLevelAt(zap.ErrorLevel))
> > > > > > > > })))
> > > > > > > > ```
> > > > > > > > 
> > > > > > > > 
> > > > > > > >     
> > > > > > > >       
> > > > > > > >     
> > > > > > > > 
> > > > > > > >       
> > > > > > > >     
> > > > > > > > 
> > > > > > > >     
> > > > > > > >   
> > > > > > > > maybe we could use zap.WrapCore custom our zapcore?
> > > > > > > 
> > > > > > > 
> > > > > > > `WithZapOptions` only affect struct `config` so the returned `zapcore.Core` is useless
> > > > > > 
> > > > > > 
> > > > > > WrapCore will take effect finally, and will cause setOutput and setLevel not to work
> > > > > 
> > > > > 
> > > > > In this case, to solve the warped `zapcore.Core` probles, we still need change methods `SetLevel` and `SetOutput` by `NewTee` and it will just change the first `zapcore.Core` in the warped `zapcore.Core`
> > > > 
> > > > 
> > > > It will look weird, and it's not so easy to change the wrapCore behavior, maybe we should avoid using these two func when using wrapCore
> > > 
> > > 
> > > ... how's weeird?
> > 
> > 
> > Why just change the first...
> 
> As It can't specify the certain `zapcore.Core`, we have to to implement `Control` interface (just point this case in docs is ok)

sounds ok

**Comment by rogerogers** [comment]
Do you join our lark community group or wechat group. contact me.

**Comment by StellarisW** [comment]
> Do you join our lark community group or wechat group. contact me.

你飞书名字是啥

**Comment by rogerogers** [comment]
> > Do you join our lark community group or wechat group. contact me.
> 
> 你飞书名字是啥

rogerogers

File: zap/option.go [hunk 1]
```diff
@@ -31,60 +31,45 @@ func (fn option) apply(cfg *config) {
...
-func WithCoreEnc(enc zapcore.Encoder) Option {
+// WithCore zapcore
+func WithCore(coreConfigs ...CoreConfig) Option {
 	return option(func(cfg *config) {
-		cfg.coreConfig.enc = enc
-	})
-}
-
-// WithCoreWs zapcore write syncer
-func WithCoreWs(ws zapcore.WriteSyncer) Option {
```
Comment (rogerogers) [inline zap/option.go:78]: Keep these methods for compatibility

File: zap/go.mod [hunk 2]
```diff
@@ -3,7 +3,10 @@ module github.com/hertz-contrib/logger/zap
 go 1.16
 
 require (
+	github.com/BurntSushi/toml v1.2.1 // indirect
 	github.com/cloudwego/hertz v0.3.2
+	github.com/natefinch/lumberjack v2.0.0+incompatible
```
Comment (rogerogers) [inline zap/go.mod]: these import is not necessary

File: zap/README.md [hunk 3]
```diff
@@ -38,4 +38,118 @@ func main() {
...
+				Enc: zapcore.NewConsoleEncoder(humaEncoderConfig()),
+				Ws:  os.Stdout,
+				Lvl: dynamicLevel,
+			},
+			{
+				Enc: zapcore.NewJSONEncoder(humanEncoderConfig()),
+				Ws:  getWriteSyncer("./all/log.log"),
+				Lvl: zap.NewAtomicLevelAt(zapcore.DebugLevel),
+			},
+			{
```
Comment (rogerogers) [inline zap/README.md:72]: One or several doesn't seem to make much sense

File: zap/option.go [hunk 4]
```diff
@@ -31,60 +31,45 @@ func (fn option) apply(cfg *config) {
...
-		zapOpts:    []zap.Option{},
+		coreConfigs: []CoreConfig{*defaultCoreConfig()},
+		zapOpts:     []zap.Option{},
 	}
 }
 
-// WithCoreEnc zapcore encoder
-func WithCoreEnc(enc zapcore.Encoder) Option {
+// WithCore zapcore
+func WithCore(coreConfigs ...CoreConfig) Option {
```
Comment (rogerogers) [inline zap/option.go]: with core use coreConfigs as param, not so clear.

File: zap/logger_test.go [hunk 5]
```diff
@@ -179,15 +192,58 @@ func TestLogLevel(t *testing.T) {
 func TestCoreOption(t *testing.T) {
 	buf := new(bytes.Buffer)
 
+	dynamicLevel := zap.NewAtomicLevel()
```
Comment (rogerogers) [inline zap/logger_test.go:234]: SetOutput，SetLevel in multi-core situations should be covered

File: zap/logger_test.go [hunk 6]
```diff
@@ -179,15 +192,58 @@ func TestLogLevel(t *testing.T) {
...
+
+	dynamicLevel.SetLevel(zap.InfoLevel)
+
 	logger := NewLogger(
-		WithCoreEnc(zapcore.NewConsoleEncoder(humanEncoderConfig())),
```
Comment (rogerogers) [inline zap/logger_test.go:183]: Add a new test and keep this?

File: zap/logger_test.go [hunk 7]
```diff
@@ -197,6 +294,11 @@ func TestCoreOption(t *testing.T) {
 	assert.True(t, strings.Contains(buf.String(), "this is a warn log"))
 	// test console encoder result
 	assert.True(t, strings.Contains(buf.String(), "\tERROR\t"))
+
+	logger.SetLevel(hlog.LevelWarn)
```
Comment (rogerogers) [inline zap/logger_test.go]: for this situation level should from high -> low, here info -> warn, error will always logged, and SetOutput not covered

**Review by rogerogers — APPROVED** [review]
LGTM

**Pull request merged**